sql: add query max memory usage to TraceAnalyzer

cathymw opened the pull request:
The first commit does a refactor of some TraceAnalyzer code.
The second commit adds max memory usage to TraceAnalyzer.
Please see individual commit messages for more details.

Closes: #54340

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56300)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cathymw)

___
*[pkg/sql/instrumentation.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MLoJcZV3qZy1C4bvxmE:-MLoJcZV3qZy1C4bvxmF:bj0wxsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfb1c4554c6b10599fb47955bf37d390f29a84f7/pkg/sql/instrumentation.go#L182)):*
> ```Go
> 		networkBytesSent := int64(0)
> 		queryMaxMem := int64(0)
> 		for _, flowInfo := range p.curPlan.distSQLFlowInfos {
> ```

Feels like this stuff should all be done inside `traceanalyzer` (or a new type in that package). There should be an interface that allows us to pass in multiple traces and flow maps and then get all the top-level stats in a struct.
___
*[pkg/sql/instrumentation.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MLoLPac7mh047xLUSWE:-MLoLPac7mh047xLUSWF:b1zo561) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfb1c4554c6b10599fb47955bf37d390f29a84f7/pkg/sql/instrumentation.go#L183)):*
> ```Go
> 		queryMaxMem := int64(0)
> 		for _, flowInfo := range p.curPlan.distSQLFlowInfos {
> 			analyzer := flowInfo.analyzer
> ```

It feels strange to store an analyzer in the struct. We should store the flow map and only initialize an analyzer when we analyze.

If we really want to not keep the flow map around, we can extract a separate `FlowMetadata` type from the `TraceAnalyzer` that is immutable and stores just the metadata (I realize TraceAnalyzer is already pretty much just that, it's more about the naming and how we describe these objects).


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cathymw and @RaduBerinde)

___
*[pkg/sql/instrumentation.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MLoLPac7mh047xLUSWE:-MMofE2y-1GM2pf4mxao:b5v186t) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfb1c4554c6b10599fb47955bf37d390f29a84f7/pkg/sql/instrumentation.go#L183)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It feels strange to store an analyzer in the struct. We should store the flow map and only initialize an analyzer when we analyze.

If we really want to not keep the flow map around, we can extract a separate `FlowMetadata` type from the `TraceAnalyzer` that is immutable and stores just the metadata (I realize TraceAnalyzer is already pretty much just that, it's more about the naming and how we describe these objects).
</blockquote></details>

Unfortunately we can't store the flow map because the lifetime doesn't extend to this piece of code (we release flow specs for allocation performance). I like the `FlowMetadata` idea.
___
*[pkg/sql/plan.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MMofc_Y4zZFeWwODJsM:-MMofc_Y4zZFeWwODJsN:by97npu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52e6b4171c9a1090698fc917ee4ebbf8f9b4f89/pkg/sql/plan.go#L316)):*
> ```Go
> 
> // GetQueryStats is part of the QueryLevelStatsGetter interface.
> func (pt *planTop) GetQueryStats(
> ```

I think it'd be nicer to have a separate type in the `execstats` package or some helper function and keep this calculation separate from the `planTop` object.
___
*[pkg/sql/colflow/vectorized_flow.go, line 398 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MMognk3-bzCLrC-V4xI:-MMognk3-bzCLrC-V4xJ:bstxkvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52e6b4171c9a1090698fc917ee4ebbf8f9b4f89/pkg/sql/colflow/vectorized_flow.go#L398)):*
> ```Go
> 	deterministicStats bool,
> 	vectorizedStatsCollectors []colexec.VectorizedStatsCollector,
> 	maxMemUsage int64,
> ```

I don't think this belongs here. It would be better to have a call after `finishVectorizedStatsCollectors` that adds this top-level flow information to the span. But I think we still need to do some work to send this message exactly once with the last outbox. I think what might be nice is to write a metadata source that is aware of the number of outboxes:
```
type FlowStatsReporter struct {
    numOutboxes int32
}

func (r *FlowStatsReporter) AddOutbox() {
    atomic.AddInt32(&r.numOutboxes, 1)
}

func (r *FlowStatsReporter) DrainMeta(ctx context.Context) []ProducerMetadata {
    if atomic.AddInt32(&r.numOutboxes, -1) == 0 {
        return []ProducerMetadata{FlowMaxMemory: r.flowCtx.Mon.GetMaxMemoryUsage()}
    }
    return nil
}
```
And then the outbox would check that this FlowMaxMemory field is set and add it to its span stats or something like this. The crucial part is that it has to be the *last* remote component because otherwise the `Mon.GetMaxMemoryUsage` might be incorrect. Keep in mind that we'll also need something similar in the row execution engine.
___
*[pkg/sql/rowflow/routers.go, line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MMog-GF60F-QBMMCSN6:-MMog-GF60F-QBMMCSN7:bawry48) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52e6b4171c9a1090698fc917ee4ebbf8f9b4f89/pkg/sql/rowflow/routers.go#L378)):*
> ```Go
> 						ro.stats.Exec.MaxAllocatedMem.Set(uint64(ro.memoryMonitor.MaximumBytes()))
> 						ro.stats.Exec.MaxAllocatedDisk.Set(uint64(ro.diskMonitor.MaximumBytes()))
> 						ro.stats.FlowStats.MaxMemUsage.Set(uint64(rb.parent.MaximumBytes()))
> ```

Is there any reason why you went down the route of sending flowstats multiple times? In the traceanalyzer, I think the current approach will be to pick the latest value received, which means that you might send a value before the flow ends (e.g. imagine this router's input tree is done before the rest of the components running in this node) and for some reason receive it last in the trace analyzer. This will report an incorrect value for memory usage. I would prefer to find a way to send this stat exactly once at the end of the flow.


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/instrumentation.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MLoJcZV3qZy1C4bvxmE:-MN_ssRsEiTkD11zvygr:bdy3v4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfb1c4554c6b10599fb47955bf37d390f29a84f7/pkg/sql/instrumentation.go#L182)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Feels like this stuff should all be done inside `traceanalyzer` (or a new type in that package). There should be an interface that allows us to pass in multiple traces and flow maps and then get all the top-level stats in a struct.
</blockquote></details>

Done. But to prevent an import cycle, I couldn't pass in `[]flowInfo`, so I get the `flowMetadata` from each of the `flowInfo`s and pass in a slice of the flowMetadata. I don't know if that's too sketchy though
___
*[pkg/sql/instrumentation.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MLoLPac7mh047xLUSWE:-MN_tr6i5sPDsNZzXsGy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfb1c4554c6b10599fb47955bf37d390f29a84f7/pkg/sql/instrumentation.go#L183)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Unfortunately we can't store the flow map because the lifetime doesn't extend to this piece of code (we release flow specs for allocation performance). I like the `FlowMetadata` idea.
</blockquote></details>

Done.
___
*[pkg/sql/plan.go, line 316 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MMofc_Y4zZFeWwODJsM:-MN_tsuR-FaiuuHPjC84:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52e6b4171c9a1090698fc917ee4ebbf8f9b4f89/pkg/sql/plan.go#L316)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it'd be nicer to have a separate type in the `execstats` package or some helper function and keep this calculation separate from the `planTop` object.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 398 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MMognk3-bzCLrC-V4xI:-MN_Ozjb1jyoZYj6ZO4L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52e6b4171c9a1090698fc917ee4ebbf8f9b4f89/pkg/sql/colflow/vectorized_flow.go#L398)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think this belongs here. It would be better to have a call after `finishVectorizedStatsCollectors` that adds this top-level flow information to the span. But I think we still need to do some work to send this message exactly once with the last outbox. I think what might be nice is to write a metadata source that is aware of the number of outboxes:
```
type FlowStatsReporter struct {
    numOutboxes int32
}

func (r *FlowStatsReporter) AddOutbox() {
    atomic.AddInt32(&r.numOutboxes, 1)
}

func (r *FlowStatsReporter) DrainMeta(ctx context.Context) []ProducerMetadata {
    if atomic.AddInt32(&r.numOutboxes, -1) == 0 {
        return []ProducerMetadata{FlowMaxMemory: r.flowCtx.Mon.GetMaxMemoryUsage()}
    }
    return nil
}
```
And then the outbox would check that this FlowMaxMemory field is set and add it to its span stats or something like this. The crucial part is that it has to be the *last* remote component because otherwise the `Mon.GetMaxMemoryUsage` might be incorrect. Keep in mind that we'll also need something similar in the row execution engine.
</blockquote></details>

Done.
___
*[pkg/sql/rowflow/routers.go, line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MMog-GF60F-QBMMCSN6:-MN_P-Ny90uB8CgT6vzw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52e6b4171c9a1090698fc917ee4ebbf8f9b4f89/pkg/sql/rowflow/routers.go#L378)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is there any reason why you went down the route of sending flowstats multiple times? In the traceanalyzer, I think the current approach will be to pick the latest value received, which means that you might send a value before the flow ends (e.g. imagine this router's input tree is done before the rest of the components running in this node) and for some reason receive it last in the trace analyzer. This will report an incorrect value for memory usage. I would prefer to find a way to send this stat exactly once at the end of the flow.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/instrumentation.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MLoJcZV3qZy1C4bvxmE:-MN_u5e99JrXF8UA0b8J:b-sxj58l) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfb1c4554c6b10599fb47955bf37d390f29a84f7/pkg/sql/instrumentation.go#L182)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

Done. But to prevent an import cycle, I couldn't pass in `[]flowInfo`, so I get the `flowMetadata` from each of the `flowInfo`s and pass in a slice of the flowMetadata. I don't know if that's too sketchy though
</blockquote></details>

^^ at `sql/instrumentation.go:190`


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> Reviewed 15 of 24 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/instrumentation.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh5eUy1UFEM42ymEig:-MNh5eUy1UFEM42ymEih:b-xkin1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/instrumentation.go#L194)):*
> ```Go
> 			flowMetadata = append(flowMetadata, flowInfo.flowMetadata)
> 		}
> 		queryLevelStats := execstats.GetQueryLevelStats(ctx, trace, cfg.TestingKnobs.DeterministicExplainAnalyze, ast, flowMetadata)
> ```

nit: don't make logging `GetQueryLevelStats`' responsibility. Change it to return an error and have callers decide what to do with it (in this case, log it). It also removes the need to pass in the `ast` just for logging purposes.
___
*[pkg/sql/plan.go, line 271 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh5NKHCYAcuSQFJrUf:-MNh5NKHCYAcuSQFJrUg:bmjmk96) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/plan.go#L271)):*
> ```Go
> 	diagram execinfrapb.FlowDiagram
> 	// FlowMetadata stores metadata from flows that will be used by TraceAnalyzer.
> 	flowMetadata *execstats.FlowMetadata
> ```

nit: I think this refactor would've been nicer separated into an earlier commit.
___
*[pkg/sql/colflow/vectorized_flow.go, line 521 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh712T9k5QNLV_Ak5z:-MNh712T9k5QNLV_Ak6-:b-rpcreq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/colflow/vectorized_flow.go#L521)):*
> ```Go
> 	// When there is one outbox, the flow max memory usage is added to a flow
> 	// level span.
> 	currOutboxes int32
> ```

It's unfortunate that we now have `numOutboxes` and `currOutboxes` that essentially indicate the same thing but they're incremented and decremented separately. I understand the need for a separate counter, but maybe we can change it so that instead of incrementing `numOutboxes` and decrementing at plan time we would have a `numOutboxes` field incremented at plan time as it currently is and then separately increment two counters, on would be `numOutboxesExited` incremented where `numOutboxes` is currently decremented, and the other would be `numOutboxesDrained` which would be incremented where `currOutboxes` is currently decremented. We would then just check the atomic and perform whatever operation we need to once it reaches `numOutboxes`. I think this is cleaner than incrementing two counters that overlap semantically.
___
*[pkg/sql/colflow/vectorized_flow.go, line 987 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhAF401oC1391Pc3SW:-MNhAF401oC1391Pc3SX:b-kprrn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/colflow/vectorized_flow.go#L987)):*
> ```Go
> 						if atomic.AddInt32(&s.currOutboxes, -1) == 0 {
> 							span.SetTag(execinfrapb.FlowIDTagKey, flowCtx.ID)
> 							span.SetSpanStats(&execinfrapb.ComponentStats{FlowStats: execinfrapb.FlowStats{MaxMemUsage: optional.MakeUint(uint64(flowCtx.EvalCtx.Mon.MaximumBytes()))}})
> ```

nit: I think this would be nicer on a couple of lines, it's a bit hard to read currently.
___
*[pkg/sql/colflow/vectorized_flow.go, line 990 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhACE0C1cIeKAuyRvi:-MNhACE0C1cIeKAuyRvj:boprrgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/colflow/vectorized_flow.go#L990)):*
> ```Go
> 						}
> 						finishVectorizedStatsCollectors(
> 							ctx, flowCtx.ID, vscs,
> ```

nit: I don't think this needs to be on a new line
___
*[pkg/sql/execstats/traceanalyzer.go, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhAbtq45YTmzij7RL4:-MNhAbtq45YTmzij7RL5:b-a7v8es) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L56)):*
> ```Go
> 	// flowStats maps a flow ID, represented as a string, to flow level stats
> 	// extracted from a trace.
> 	flowStats map[string]*flowStats
> ```

Why is this a string and not a `FlowID`?
___
*[pkg/sql/execstats/traceanalyzer.go, line 100 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhB-IdB0kgk_Uw2I-b:-MNhB-IdB0kgk_Uw2I-c:b-4ds4t0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L100)):*
> ```Go
> // NewTraceAnalyzer creates a TraceAnalyzer with the corresponding physical
> // plan. Call AddTrace to calculate meaningful stats.
> func NewTraceAnalyzer(flowMetadata FlowMetadata) *TraceAnalyzer {
> ```

I believe this can be changed to `MakeTraceAnalyzer` which would return a `TraceAnalyzer` value (no pointer) if you embed `FlowMetadata` as `*FlowMetadata` then you could also remove change the pointer receiver in `AddTrace` to a value receiver.
___
*[pkg/sql/execstats/traceanalyzer.go, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhBcIy4v6TipSodHpX:-MNhBcIy4v6TipSodHpY:bo3sy29) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L151)):*
> ```Go
> 				return errors.Errorf("trace has span for stream %d but the stream does not exist in the physical plan", id)
> 			}
> 			streamStats.stats = append(streamStats.stats, &stats)
> ```

This should no longer be necessary, right?
___
*[pkg/sql/execstats/traceanalyzer.go, line 181 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhBLab1kfns5BD3rdh:-MNhBLab1kfns5BD3rdi:bdtb2ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L181)):*
> ```Go
> 	if v.FlowStats.MaxMemUsage.HasValue() {
> 		// This stream might have multiple span stats, some of which may not have network bytes information.
> 		return 0, nil
> ```

How are we hitting this case? In colexec we set up stats with only a flow id tag so it can't be confused for stream stats and in rowexec we set the `BytesSent` value which would mean we exit above. 
___
*[pkg/sql/execstats/traceanalyzer.go, line 255 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhBzH7ENT7kla_3IRw:-MNhBzH7ENT7kla_3IRx:b-33xlb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L255)):*
> ```Go
> 	// TODO(cathymw): maxMemUsage shouldn't be attached to span stats that are associated with streams,
> 	// since it's a flow level stat. However, due to the row exec engine infrastructure, it is too
> 	// complicated to attach this to a flow level span. If the row exec engine gets removed, getting
> ```

This comment should also be in the rowexec outbox when we set the stats.
___
*[pkg/sql/execstats/traceanalyzer.go, line 276 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhC8KI5ozJ0rCiKbnk:-MNhC8KI5ozJ0rCiKbnl:bl6wxk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L276)):*
> ```Go
> 	ast tree.Statement,
> 	flowMetadata []*FlowMetadata,
> ) *QueryLevelStats {
> ```

I don't think `QueryLevelStats` needs to be a pointer.
___
*[pkg/sql/flowinfra/outbox.go, line 84 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh3y-_B9daQr4xXFAs:-MNh3y-_B9daQr4xXFAt:b-my2zg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/flowinfra/outbox.go#L84)):*
> ```Go
> 	flowID execinfrapb.FlowID,
> 	streamID execinfrapb.StreamID,
> 	numOutboxes int32,
> ```

I don't know how this is working. Passing in an `int32` by value will make a copy of the value, so each outbox will independently decrement their own value. Only the first outbox created will end up setting stats because `numOutboxes` was `1` at its time of creation. You should probably pass in a pointer instead.


<!-- Sent from Reviewable.io -->

cathymw committed:
sql: refactor TraceAnalyzer

This commit introduces a new type FlowMetadata to avoid flowInfo from storing
a TraceAnalyzer. This commit also moves code from instrumentation.go to a
helper function GetQueryLevelStats(), which takes in trace and flow metadata
and returns top level stats.

Release note: None.

cathymw committed:
sql: add query max memory usage to TraceAnalyzer

This commit adds query max memory usage to TraceAnalyzer, which
will later be surfaced to the statement stats page on the
admin ui. This approach uses the parent memory monitor to accurately
report the max memory that is being used at any point in the query.

Release note: None.

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/instrumentation.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh5eUy1UFEM42ymEig:-MNisR8IEnwwnI1odw0H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/instrumentation.go#L194)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: don't make logging `GetQueryLevelStats`' responsibility. Change it to return an error and have callers decide what to do with it (in this case, log it). It also removes the need to pass in the `ast` just for logging purposes.
</blockquote></details>

Done.
___
*[pkg/sql/plan.go, line 271 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh5NKHCYAcuSQFJrUf:-MNjugtlBssp_--WDZzM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/plan.go#L271)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I think this refactor would've been nicer separated into an earlier commit.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 521 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh712T9k5QNLV_Ak5z:-MNj7DH4ASjyBZVWwQKi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/colflow/vectorized_flow.go#L521)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It's unfortunate that we now have `numOutboxes` and `currOutboxes` that essentially indicate the same thing but they're incremented and decremented separately. I understand the need for a separate counter, but maybe we can change it so that instead of incrementing `numOutboxes` and decrementing at plan time we would have a `numOutboxes` field incremented at plan time as it currently is and then separately increment two counters, on would be `numOutboxesExited` incremented where `numOutboxes` is currently decremented, and the other would be `numOutboxesDrained` which would be incremented where `currOutboxes` is currently decremented. We would then just check the atomic and perform whatever operation we need to once it reaches `numOutboxes`. I think this is cleaner than incrementing two counters that overlap semantically.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 987 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhAF401oC1391Pc3SW:-MNj3Bdq2Fl1xYJFvcjz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/colflow/vectorized_flow.go#L987)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I think this would be nicer on a couple of lines, it's a bit hard to read currently.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 990 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhACE0C1cIeKAuyRvi:-MNj3B3-8_75q5MbbYcM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/colflow/vectorized_flow.go#L990)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I don't think this needs to be on a new line
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 56 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhAbtq45YTmzij7RL4:-MNj0cGW8ZBjk4lqP4io:b-cumfrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L56)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why is this a string and not a `FlowID`?
</blockquote></details>

Done. Changed to FlowID.
___
*[pkg/sql/execstats/traceanalyzer.go, line 100 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhB-IdB0kgk_Uw2I-b:-MNixJIG3ch_RJOw60K0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L100)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I believe this can be changed to `MakeTraceAnalyzer` which would return a `TraceAnalyzer` value (no pointer) if you embed `FlowMetadata` as `*FlowMetadata` then you could also remove change the pointer receiver in `AddTrace` to a value receiver.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhBcIy4v6TipSodHpX:-MNiwNKT3q6eVsmRYsJJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L151)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This should no longer be necessary, right?
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 181 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhBLab1kfns5BD3rdh:-MNiwEgoEe4C46op8v3d:b-b1pcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L181)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

How are we hitting this case? In colexec we set up stats with only a flow id tag so it can't be confused for stream stats and in rowexec we set the `BytesSent` value which would mean we exit above. 
</blockquote></details>

Done. Oops, you're right, removed.
___
*[pkg/sql/execstats/traceanalyzer.go, line 255 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhBzH7ENT7kla_3IRw:-MNiutq-1tfQjPC0Aa3e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L255)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This comment should also be in the rowexec outbox when we set the stats.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 276 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNhC8KI5ozJ0rCiKbnk:-MNisUhX9PgIzYrtfe7n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/execstats/traceanalyzer.go#L276)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think `QueryLevelStats` needs to be a pointer.
</blockquote></details>

Done.
___
*[pkg/sql/flowinfra/outbox.go, line 84 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh3y-_B9daQr4xXFAs:-MNiuV66EZrgOwsYgo7w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/flowinfra/outbox.go#L84)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't know how this is working. Passing in an `int32` by value will make a copy of the value, so each outbox will independently decrement their own value. Only the first outbox created will end up setting stats because `numOutboxes` was `1` at its time of creation. You should probably pass in a pointer instead.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> Reviewed 1 of 24 files at r3, 13 of 13 files at r4, 11 of 11 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, and @yuzefovich)

___
*[pkg/sql/instrumentation.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh5eUy1UFEM42ymEig:-MNwYYN1ACEn4o0E1fQw:bbcmi26) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/instrumentation.go#L194)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: `errorMsgs` is redundant with `errors` since the latter already contains a message. Oh, I saw you've amended this in the latter commit after I wrote this message, but that should probably be in the first commit.
___
*[pkg/sql/colflow/vectorized_flow.go, line 495 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwbRqi126MS3LYYHM-:-MNwbRqi126MS3LYYHM0:b-6r56vb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/colflow/vectorized_flow.go#L495)):*
> ```Go
> 	numOutboxesExited int32
> 	// numOutboxesDrained is an atomic that keeps track of how many outboxes have
> 	// been drained. When numOutboxesDrained equals numOutboxes, the flow max memory
> ```

nit: it doesn't hurt to be more general to avoid this comment becoming stale in the future something like `s/the flow max memory usage is added to a flow level span/flow-level metadata is sent`.
___
*[pkg/sql/colflow/vectorized_flow.go, line 988 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwbqR5CiLf9Z4_R403:-MNwbqR5CiLf9Z4_R404:bk9z0so) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/colflow/vectorized_flow.go#L988)):*
> ```Go
> 						ctx, span := tracing.ChildSpanRemote(ctx, "")
> 						if atomic.AddInt32(&s.numOutboxesDrained, 1) == atomic.LoadInt32(&s.numOutboxes) {
> 							span.SetTag(execinfrapb.FlowIDTagKey, flowCtx.ID)
> ```

Add a comment about why the last outbox is doing this.
___
*[pkg/sql/execstats/traceanalyzer.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwYN82-0r9raOWNfjx:-MNwYN82-0r9raOWNfjy:bw3wj6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3c3c8156ffd912f0bbcafbf5c8206a8ba3311a/pkg/sql/execstats/traceanalyzer.go#L52)):*
> ```Go
> 	// for the given stream in the trace.
> 	streamStats map[execinfrapb.StreamID]*streamStats
> 	// flowStats maps a flow ID, represented as a string, to flow level stats
> ```

nit: `represented as a string` is no longer true
___
*[pkg/sql/execstats/traceanalyzer.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwZdh98DxF0xmbhhY9:-MNwZdh98DxF0xmbhhYA:b-shm0ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3c3c8156ffd912f0bbcafbf5c8206a8ba3311a/pkg/sql/execstats/traceanalyzer.go#L206)):*
> ```Go
> }
> 
> // GetQueryLevelStats returns all the top-level stats in a QueryLevelStats struct.
> ```

nit: explain the multiple `[]error` behavior since it's unusual. I think the important point is that `GetQueryLevelStats` is best-effort in that it tries to calculate as many stats as possible.
___
*[pkg/sql/execstats/traceanalyzer.go, line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwZts80Mg2HQdi2Olq:-MNwZts80Mg2HQdi2Olr:b-5rwx47) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3c3c8156ffd912f0bbcafbf5c8206a8ba3311a/pkg/sql/execstats/traceanalyzer.go#L225)):*
> ```Go
> 		if err != nil {
> 			errors = append(errors, err)
> 			errorMsgs = append(errorMsgs, "error calculating network bytes sent for stmt %s: %v")
> ```

Don't you need to `continue` here?
___
*[pkg/sql/execstats/traceanalyzer.go, line 268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcCYW52Q8q5YP204o:-MNwcCYW52Q8q5YP204p:b-asf1cq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/execstats/traceanalyzer.go#L268)):*
> ```Go
> 		analyzer := MakeTraceAnalyzer(metadata)
> 		if err := analyzer.AddTrace(trace, deterministicExplainAnalyze); err != nil {
> 			errors = append(errors, err)
> ```

It might be nice to still `Wrap` this error with some way to identify which stat produced an error (in the first commit).
___
*[pkg/sql/flowinfra/outbox.go, line 69 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcYax2NwI21Cqa5D4:-MNwcYax2NwI21Cqa5D5:bkh5i7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/flowinfra/outbox.go#L69)):*
> ```Go
> 	stats                  execinfrapb.ComponentStats
> 
> 	// currOutboxes is an atomic that keeps track of how many outboxes are left.
> ```

nit: `s/currOutboxes/numOutboxes`
___
*[pkg/sql/flowinfra/outbox_test.go, line 77 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcj5vA7TyWAkuGiCS:-MNwcj5vA7TyWAkuGiCT:b-e2z0os) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/flowinfra/outbox_test.go#L77)):*
> ```Go
> 	streamID := execinfrapb.StreamID(42)
> 	numOutboxes := int32(0)
> 	outbox := NewOutbox(&flowCtx, execinfra.StaticNodeID, flowID, streamID, &numOutboxes)
> ```

nit: why not just pass in `nil` and only check `m.numOutboxes` in `flowinfra/outbox.go` if the pointer is non-nil.
___
*[pkg/sql/rowflow/row_based_flow.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcwlRDWXw-EMu7YL4:-MNwcwlRDWXw-EMu7YL5:b-3qdpqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/rowflow/row_based_flow.go#L34)):*
> ```Go
> 
> 	// currOutboxes is an atomic that counts how many outboxes have been created.
> 	// This atomic is then used by the last outbox to know when to report the flow's
> ```

nit: same comment regarding making this comment more general. Also mention somewhere in the rowflow code *why* it's necessary for the last outbox to do this.
___
*[pkg/sql/rowflow/row_based_flow.go, line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcrty7tVwhe4XVm05:-MNwcrty7tVwhe4XVm06:b5uaohr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/rowflow/row_based_flow.go#L36)):*
> ```Go
> 	// This atomic is then used by the last outbox to know when to report the flow's
> 	// max memory usage.
> 	currOutboxes int32
> ```

nit: rename this to `numOutboxes` since that's what's used everywhere else.


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/instrumentation.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNh5eUy1UFEM42ymEig:-MNy4p2X9ETu0CU7CZB_:bt67mbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46bbcf09bf86eb4a058c5aaea882c8dbf3fb2c2/pkg/sql/instrumentation.go#L194)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `errorMsgs` is redundant with `errors` since the latter already contains a message. Oh, I saw you've amended this in the latter commit after I wrote this message, but that should probably be in the first commit.
</blockquote></details>

Done. Moved to first commit.
___
*[pkg/sql/colflow/vectorized_flow.go, line 495 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwbRqi126MS3LYYHM-:-MNy7IhCDdF6pMg0gIgI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/colflow/vectorized_flow.go#L495)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: it doesn't hurt to be more general to avoid this comment becoming stale in the future something like `s/the flow max memory usage is added to a flow level span/flow-level metadata is sent`.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 988 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwbqR5CiLf9Z4_R403:-MNyg4YS87Gt6m1sCoO6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/colflow/vectorized_flow.go#L988)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Add a comment about why the last outbox is doing this.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwYN82-0r9raOWNfjx:-MNy54KM6ShUZmJqfGvY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3c3c8156ffd912f0bbcafbf5c8206a8ba3311a/pkg/sql/execstats/traceanalyzer.go#L52)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `represented as a string` is no longer true
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwZdh98DxF0xmbhhY9:-MNy4iwA8Jxg8Npu7zFf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3c3c8156ffd912f0bbcafbf5c8206a8ba3311a/pkg/sql/execstats/traceanalyzer.go#L206)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: explain the multiple `[]error` behavior since it's unusual. I think the important point is that `GetQueryLevelStats` is best-effort in that it tries to calculate as many stats as possible.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwZts80Mg2HQdi2Olq:-MNxte0dDaCkTHpvzDE9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3c3c8156ffd912f0bbcafbf5c8206a8ba3311a/pkg/sql/execstats/traceanalyzer.go#L225)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Don't you need to `continue` here?
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcCYW52Q8q5YP204o:-MNy2xRnEBD8WkXv4qcS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/execstats/traceanalyzer.go#L268)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It might be nice to still `Wrap` this error with some way to identify which stat produced an error (in the first commit).
</blockquote></details>

Done.
___
*[pkg/sql/flowinfra/outbox.go, line 69 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcYax2NwI21Cqa5D4:-MNygL5NCLGQ-Gr_UsDS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/flowinfra/outbox.go#L69)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: `s/currOutboxes/numOutboxes`
</blockquote></details>

Done.
___
*[pkg/sql/flowinfra/outbox_test.go, line 77 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcj5vA7TyWAkuGiCS:-MNyhLeH56Je8U_Kv8cF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/flowinfra/outbox_test.go#L77)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: why not just pass in `nil` and only check `m.numOutboxes` in `flowinfra/outbox.go` if the pointer is non-nil.
</blockquote></details>

Done.
___
*[pkg/sql/rowflow/row_based_flow.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcwlRDWXw-EMu7YL4:-MNyoOy37r6qDiUIgd62:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/rowflow/row_based_flow.go#L34)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: same comment regarding making this comment more general. Also mention somewhere in the rowflow code *why* it's necessary for the last outbox to do this.
</blockquote></details>

Done.
___
*[pkg/sql/rowflow/row_based_flow.go, line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcrty7tVwhe4XVm05:-MNyi6FBDbqh2LZFUlow:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/rowflow/row_based_flow.go#L36)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: rename this to `numOutboxes` since that's what's used everywhere else.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
Good job! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> minus some remaining nits

> Reviewed 10 of 10 files at r6, 10 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-:-MO0g4JS9KIbm_P15DIq:broypt6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcCYW52Q8q5YP204o:-MO0ehYTEa-2yr2E_A9c:b-k2criw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/execstats/traceanalyzer.go#L268)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: I don't think the message makes sense as written, did you mean `error analyzing`?
___
*[pkg/sql/execstats/traceanalyzer.go, line 208 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MO0eaBL3nOlbIDd1_CM:-MO0eaBL3nOlbIDd1_CN:b-xrxci8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3022fbc03c0d0eab07a7181e3158d763d067f7eb/pkg/sql/execstats/traceanalyzer.go#L208)):*
> ```Go
> // GetQueryLevelStats tries to calculate as many stats as possible. If errors occur
> // while calculating stats, GetQueryLevelStats adds the error to a slice to be returned
> // for logging purposes, but continues calculating other stats.
> ```

small nit: `GetQueryLevelStats` doesn't know what the errors will be used for (some caller in the future might decide to return the last errror) so this should not reference any logging


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MNwcCYW52Q8q5YP204o:-MO2a5u_EK7j1fho-MMp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8de619979dacaefa5125e403ed38c5d7ecd24bc/pkg/sql/execstats/traceanalyzer.go#L268)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I don't think the message makes sense as written, did you mean `error analyzing`?
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 208 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-MO0eaBL3nOlbIDd1_CM:-MO2a6aQ3AL1XZbk1KlG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3022fbc03c0d0eab07a7181e3158d763d067f7eb/pkg/sql/execstats/traceanalyzer.go#L208)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

small nit: `GetQueryLevelStats` doesn't know what the errors will be used for (some caller in the future might decide to return the last errror) so this should not reference any logging
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :shipit: 

> Reviewed 12 of 12 files at r8, 10 of 10 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56300#-:-MO5rplg9MIEmggHjoG7:byzm7aw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

cathymw commented:
TFTRs!
bors r=asubiotto
